libs/Microsoft.MixedReality.WebRTC/VideoTrackSource.cs

```diff
@@ -263,13 +274,19 @@ internal void OnTracksRemovedFromSource(IEnumerable<LocalVideoTrack> tracks)
         internal void OnI420AFrameReady(I420AVideoFrame frame)
         {
             MainEventSource.Log.I420ALocalVideoFrameReady(frame.width, frame.height);
-            _videoFrameReady?.Invoke(frame);
+            lock (_videoFrameReadyLock)
+            {
+                _videoFrameReady?.Invoke(frame);
+            }
         }
 
         internal void OnArgb32FrameReady(Argb32VideoFrame frame)
         {
             MainEventSource.Log.Argb32LocalVideoFrameReady(frame.width, frame.height);
-            _argb32videoFrameReady?.Invoke(frame);
+            lock (_videoFrameReadyLock)
+            {
+                _argb32videoFrameReady?.Invoke(frame);
+            }
```

Comment:
I prefer to leave the lock, because the first handler registered is used to determine when to register the native callback, and this is not atomic, so has more potential to cause hard-to-debug issues than a deadlock which is plain and visible. I documented instead that handlers must not add/remove handlers to the event while running.

Reply:
There has to be a lock on add/remove because of the non-atomicity as you say, I was suggesting to avoid the lock _just_ here in the callbacks. `_argb32videoFrameReady?.Invoke(frame)` should read the EventHandler atomically and thus shouldn't need to be under lock. The only issue with this is that (see linked article) if a handler is `remove`d while a call to Invoke is ongoing, the Invoke thread will hold the handler alive until the end of the function, and also call the handler, which might be surprising and/or make disposal logic a mess.

Anyway fine by me as it is right now.